openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -293,161 +585,257 @@ def to_list(self, discard_cosmetic_attributes=True):
+                    if attrib_basename in smirnoff_data:
+                        err_msg = (f"The attribute '{attrib_basename}' has been specified "
+                                   f"with and without index: '{attrib_w_index}'")
+                        raise TypeError(err_msg)
 
+                    # Otherwise create the list object.
+                    smirnoff_data[attrib_basename] = list()
 
-        # Ensure that all required attribs are present
-        for reqd_attrib in self._REQUIRED_SPEC_ATTRIBS:
-            # SMIRKS are a special case which is handled above
-            if reqd_attrib == 'smirks':
-                continue
-            if not reqd_attrib in smirnoff_data:
-                msg = "{} requires {} as a parameter during initialization, however this is not " \
-                      "provided. Defined kwargs are {}".format(self.__class__,
-                                                               reqd_attrib,
-                                                               list(smirnoff_data.keys()))
-                raise SMIRNOFFSpecError(msg)
+                # Append the new value to the list.
+                smirnoff_data[attrib_basename].append(attrib_w_index_value)
 
+                # Remove the indexed attribute from the kwargs as it will
+                # be exposed only as an element of the list.
+                del smirnoff_data[attrib_w_index]
+                index += 1
 
-        # Finally, set attributes of this ParameterType
+            # Update the lengths with this attribute (if it was found).
+            if index > 1:
+                indexed_attr_lengths[attrib_basename] = len(smirnoff_data[attrib_basename])
+
+        # Raise an error if we there are different indexed
+        # attributes with a different number of terms.
+        if len(set(indexed_attr_lengths.values())) > 1:
+            raise TypeError('The following indexed attributes have '
+                            f'different lengths: {indexed_attr_lengths}')
+
+        # Check for missing required arguments.
+        given_attributes = set(smirnoff_data.keys())
+        required_attributes = set(self._get_required_parameter_attributes().keys())
+        missing_attributes = required_attributes.difference(given_attributes)
+        if len(missing_attributes) != 0:
+            msg = (f"{self.__class__} require the following missing parameters: {sorted(missing_attributes)}."
+                   f" Defined kwargs are {sorted(smirnoff_data.keys())}")
+            raise SMIRNOFFSpecError(msg)
+
+        # Finally, set attributes of this ParameterType and handle cosmetic attributes.
+        allowed_attributes = set(self._get_parameter_attributes().keys())
         for key, val in smirnoff_data.items():
-            if key in allowed_attribs:
-                # TODO: create @property.setter here if attrib requires unit
+            if key in allowed_attributes:
                 setattr(self, key, val)
```

Comment:
Ahh, so, in this case, `setattr` will send `val` through the `ParameterAttribute`s setter validator. Very cool.